### PR TITLE
2D/3D numerical model tutorials

### PR DESCRIPTION
**Pull request by boriskaus**
This tutorial explains how to create 2D and 3D numerical model subduction setups (with and without temperature).

Addresses issue #76

### CODE REVIEW

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+nx,nz = 512,128
+x = range(-1000,1000, nx);
+z = range(-660,0,    nz);
+Grid2D = CartData(XYZGrid(x,0,z))
+
+# Now we create an integer array that will hold the `Phases` information (which usually refers to the material or rock type in the simulation)
+Phases = zeros(Int64,size(Grid2D));
+
+# In many (geodynamic) models, one also has to define the temperature, so lets initiate it
+Temp = ones(Float64,size(Grid2D))*1350;
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:26]:
```suggestion
Temp = ones(Float64,size(Grid2D))*1350;
```
`Temp = fill(1350.0, nx, 1, nz)`

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+=#
+using GeophysicalModelGenerator
+
+nx,nz = 512,128
+x = range(-1000,1000, nx);
+z = range(-660,0,    nz);
+Grid2D = CartData(XYZGrid(x,0,z))
+
+# Now we create an integer array that will hold the `Phases` information (which usually refers to the material or rock type in the simulation)
+Phases = zeros(Int64,size(Grid2D));
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:23]:
```suggestion
Phases = zeros(Int64,size(Grid2D));
```
`Phases = zeros(Int64, nx, 1, nz)` I think is a bit more transparent to the reader

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+# Add them to the `CartData` dataset:
+Grid2D = addField(Grid2D,(;Phases, Temp))
+
+# Which looks like
+Write_Paraview(Grid2D,"Grid2D_SubductionMechanical");
+# ![Mechanical2D_Tutorial_1](../assets/img/Mechanical2D_Tutorial_1.png) 
+
+# #### Add lithospheric layers
+# In many geodynamic models, the lithosphere consists of a crust and mantle (or upper crust, lower crust and mantle lithosphere).
+# We can use the function `LithosphericPhases` for this which is a simple way to set a lithospheric layering
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:45]:
```suggestion
# We can use the function `LithosphericPhases` for this, which is a simple way to set a lithospheric layering
```

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+AddBox!(Phases, Temp, Grid2D; xlim=(-800,0.0), zlim=(-80.0, 0.0), phase = lith);
+AddBox!(Phases, Temp, Grid2D; xlim=(0,300), zlim=(-80.0, 0.0), phase = lith, DipAngle=30);
+
+# Which looks like:
+Grid2D = addField(Grid2D,(;Phases, Temp))
+Write_Paraview(Grid2D,"Grid2D_SubductionMechanicalLayered");
+# ![Mechanical2D_Tutorial_2](../assets/img/Mechanical2D_Tutorial_2.png) 
+
+# #### Add halfspace cooling thermal structure
+# Sofar, we only created the mechanical structure but not the thermal part.
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:58]:
```suggestion
# So far, we only created the mechanical structure but not the thermal one.
```

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+# We can do that by specifying a thermal structure
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:59]:
```suggestion
# We can do that by specifying a thermal structure. For example, we can use the half-space cooling model:
```

File: tutorials/Tutorial_NumericalModel_2D.jl [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+therm = HalfspaceCoolingTemp(Age=40)
+AddBox!(Phases, Temp, Grid2D; xlim=(-800,0.0), zlim=(-80.0, 0.0), phase = lith, T=therm);
+AddBox!(Phases, Temp, Grid2D; xlim=(0,300), zlim=(-80.0, 0.0), phase = lith, T = therm, DipAngle=30);
+
+# Which looks like:
+Grid2D = addField(Grid2D,(;Phases, Temp))
+Write_Paraview(Grid2D,"Grid2D_SubductionHalfspaceCooling");
+# ![Mechanical2D_Tutorial_3](../assets/img/Mechanical2D_Tutorial_3.png) 
+
+# Note that you can specify several 1D thermal structures, such as
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_2D.jl:69]:
```suggestion
# Note that you can specify other 1D thermal profiles, such as 
```

**Review by albert-de-montserrat — COMMENTED** [review]
Looks good to me! I'm able to reprodce the plots withour problems

**Comment by mthielma** [comment]
Both 2D and 3D tutorial work for me.

**Comment by boriskaus** [comment]
> Both 2D and 3D tutorial work for me.

thanks for testing. I'm planning to expand the 3D tutorial to include the funky new functions of the Bayreuth team, such that this will be the last example in the JOSS paper

File: tutorials/Tutorial_NumericalModel_3D.jl [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+lith = LithosphericPhases(Layers=[15 15 45 40], Phases=[7 0 1 2], Tlab=1250)
+
+AddBox!(Phases, Temp, Grid; xlim=(200,700), ylim=(-1000, 0.0), zlim=(-200.0, 15), phase = lith,  T=ConstantTemp(T=1350), DipAngle=30);
+AddBox!(Phases, Temp, Grid; xlim=(200,700), ylim=(0, 100.0), zlim=(-200.0, 15), phase = lith,  T=ConstantTemp(T=1350), DipAngle=35);
+
+AddBox!(Phases, Temp, Grid; xlim=(200,700), ylim=(-1000, 0.0), zlim=(-110.0, 15), phase = lith,  T=McKenzie_subducting_slab(Tsurface=0,v_cm_yr=3), DipAngle=30);
+AddBox!(Phases, Temp, Grid; xlim=(200,700), ylim=(0, 1000), zlim=(-110.0, 15), phase = lith,  T=McKenzie_subducting_slab(Tsurface=0,v_cm_yr=3), DipAngle=35);
+
+
+ind=findall(isnan.(Temp)); Temp[ind] .= 0;
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_3D.jl:87]:
```suggestion
map!(x -> isnan(x) ? zero(x) : x, Temp, Temp);
```
This is very MATLAB style (and allocates few intermediate arrays), could be replace by something more Julian.

Comment (boriskaus) [reply tutorials/Tutorial_NumericalModel_3D.jl:87]: hmm, I personally find your Julia way pretty hard to understand...
Can this be done in a less cryptic way?

Comment (albert-de-montserrat) [reply tutorials/Tutorial_NumericalModel_3D.jl:87]: What about?
```julia
replace!(Temp, NaN => 0.0)
```

Comment (boriskaus) [reply tutorials/Tutorial_NumericalModel_3D.jl:87]: ok that is indeed easier. changed that accordingly

File: tutorials/Tutorial_NumericalModel_3D.jl [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+# In many (geodynamic) models, one also has to define the temperature, so lets define it as well
+Temp = fill(1350.0, nx,ny,nz);
+
+# #### Simple free subduction setup
+
+# Much of the options are explained in the 2D tutorial, which can directly be transferred to 3D.
+# Therefore, we will start with a simple subduction setup, which consists of a horizontal part that has a mid-oceanic ridge on one explained
+
+# We use a lithospheric structure. Note that if the lowermost layer has the same phase as the mantle, you can define `Tlab` as the lithosphere-asthenosphere boundary which will automatically adjust the phase depending on temperature
+lith = LithosphericPhases(Layers=[15 45 10], Phases=[0 1 2], Tlab=1250)
```
Comment (albert-de-montserrat) [inline tutorials/Tutorial_NumericalModel_3D.jl:35]: Perhaps I'm missing where this is explained, but it was not clear to me what `Layers=[15 45 10]`  means. It's the depth distance from the previous layer right?

**Review by albert-de-montserrat — APPROVED** [review]

File: tutorials/Tutorial_NumericalModel_3D.jl (hunk 8 above)
Comment (boriskaus) [reply tutorials/Tutorial_NumericalModel_3D.jl:36]: I have extended that

**Review by aelligp — APPROVED** [review]

**Pull request merged**